docs/design/speed-test.md

```diff
@@ -0,0 +1,203 @@
+these configurations to authenticate and perform uploads, measuring speed and reporting results through the CRD status.
+
+## Background
+
+We have observed several instances in customer environments where backup or restore operations became stalled or significantly delayed, but the root cause was not immediately apparent.
+These issues could stem from poor network connectivity, cloud storage throttling, or misconfigurations in BackupStorageLocation (BSL) settings. Since backup performance plays a critical
+role in disaster recovery strategies, it is essential to have a reliable way to test upload speeds.
+
+This Upload Speed Test CRD will help in such scenarios by proactively measuring the upload speed from the OpenShift cluster to the cloud storage configured in OADP’s BSL. By identifying
+performance bottlenecks ahead of backup or restore operations, administrators can take corrective actions to avoid failures during critical operations.
+
+## Goals
+
+-   Measure upload speeds for a given BackupStorageLocation (BSL) config settings and credentials
+-   Measure upload speeds from the OpenShift cluster to the configured object storage.
+-   Update the test results in the status of the UploadSpeedTest CRD for visibility and tracking.
+-   Enable smooth integration with OADP operator-managed backup storage location configurations.
+
+
+## Non-Goals
+
+- This design will not create or modify BackupStorageLocation entries in OADP.
+- It will not implement download or latency tests, focusing solely on upload speed.
+- Scheduling of recurring tests is not supported in the initial version.
+
+## High-Level Design
+Components involved and their responsibilities:
+
+- UploadSpeedTest (UST) CRD:
+  - Captures the BSL configuration, test file size and test timeout used for testing directly in the CRD.
+  - Holds the test results, including success/failure and upload speed.
+- UploadSpeedTest (UST) Controller:
+  - Watches for changes to the UploadSpeedTest CRs and executes uploads based on the provided BSL configuration. 
+  - Uses appropriate cloud SDKs (AWS, Azure, or GCP) to perform uploads.
+  - Updates the status of UST CR with upload speed, upload success/failure and error message if any.
+- OADP/DPA-UST Integration:
+  - Fetches BackupStorageLocation (BSL) details from the OADP operator’s DPA CR. 
+  - Populates the UploadSpeedTest CR with relevant configurations from the BSL.
+  - Creates the UST CR for the BSL configured in DPA instance.
+## Detailed Design
+
+#### Proposed Upload Speed Test CRD:
+
+The UploadSpeedTest CRD structure includes fields to capture backup storage location configurations, upload test parameters, and the location of cloud provider secrets.
+
+```yaml
+apiVersion: oadp.openshift.io/v1alpha1
+kind: UploadSpeedTest
+metadata:
+  name: my-upload-speed-test
+spec:
+  backupLocation:
```

Comment:
```suggestion
    backupStorageLocationName: <name of bsl to test>
    backupLocation:
```

Reply:
I think we want UST to be able to be used independently (in addition to via DPA) by providing the BSLConfig. Thinking of the the use-case:
- User has a BSL Config,
- they test the BSLConfig via UST
- now use the BSLConfig to create BSL

Reply:
So we do not want to reference by name to an already existing BSL?

Reply:
Yeah because we already have the functionality to test a particular BSL via DPA, [proposed here](https://github.com/openshift/oadp-operator/blob/7250bac87d17dcc25417ee87b8644906bda24523/docs/design/speed-test.md?plain=1#L91) . Why keep it bi-directional ! I believe this keeps the CRD simpler.

Reply:
So an upstream velero user would not be able to use this without copying the whole bsl.spec?

Reply:
Perhaps that was not the plan.

Reply:
not an upstream feature :)